src/transformStateWithClones.js

```diff
@@ -6,8 +6,48 @@
  *
  * @return {Object[]}
  */
+function copyStateAndPush(Obj, arr) {
+  const copyObj = { ...Obj };
+
+  arr.push(copyObj);
+}
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const historyArray = [];
+  const copyState = { ...state };
+
+  for (const act in actions) {
+    const typeAction = Object.values(actions[act]);
+    const commandData = typeAction[1];
+
+    if (typeAction[0] === 'addProperties') {
```

Comment:
You should simply use `actions[act].type` for the type and `actions[act].extraData` (or `actions[act].keysToRemove` if type is `removeProperties`) this is far more readable. Also your current implementation depends on the order of definition of the objects properties